src/base/bittorrent/session.cpp

```diff
@@ -1929,13 +1929,21 @@ bool Session::addTorrent_impl(CreateTorrentParams params, const MagnetUri &magne
                 , (patchedFastresumeData.constData() + patchedFastresumeData.size())};
             p.flags |= lt::add_torrent_params::flag_use_resume_save_path;
 
+            // load from .torrent file when fastresume doesn't contain the required `info` dict
+            if (!patchedFastresumeData.contains("info-hash20:"))
```

Comment:
Are you sure you should check for `info-hash` field?

Reply:
Also you still load .torrent file unconditionally in `startupTorrents()` and pass it in `addTorrent_impl()`. I think the more deep changes are required.

Reply:
>Are you sure you should check for info-hash field?

I think it should be more unique than a `info` dict.

>Also you still load .torrent file unconditionally in startupTorrents() and pass it in addTorrent_impl(). I think the more deep changes are required.

Do you insist on that? Following your plan we should be able to drop loading .torrent file in a later qbt release without any hassle. If we do it now, (I suspect) the related code might scatter and makes it harder to maintain (besides the fact that I don't really want to touch the complex `startupTorrents()`).
Maybe just add a TODO comment in `startupTorrents()` that will mention this plan?

Reply:
>I think it should be more unique than a `info` dict.

I'm sorry, I can't figure out how it's connected. You have to check that `info` exists in resume data. Why do you check other field?

Reply:
>Following your plan we should be able to drop loading .torrent file in a later qbt release without any hassle. If we do it now, (I suspect) the related code might scatter and makes it harder to maintain

But the changes will be the same no matter when we do it. Why do you think it should be harder now than later?

>besides the fact that I don't really want to touch the complex startupTorrents()

That's a plausible reason.
But we still have to do it sooner or later, if we want our program does not do unnecessary actions, and our code does not cause reasonable questions about its logic.

>Maybe just add a TODO comment in `startupTorrents()` that will mention this plan?

Yes, at least.

Reply:
>You have to check that info exists in resume data. Why do you check other field?

I made a mistake about the relations of `info` and `info-hash`, fixed now.

Reply:
>Why do you think it should be harder now than later?

Because you only need to strip out code instead of stuffing more logic in.

>That's a plausible reason.

Have you considered how it would be done? Or you would like to provide some help/code on this one?

>>Maybe just add a TODO comment in startupTorrents() that will mention this plan?
>
>Yes, at least.

Added.

Reply:
>Because you only need to strip out code instead of stuffing more logic in.

Well, I see. You expect all users to go through some intermediate version (e.g. 4.2.x) and they will have a resume data in updated format (with info dict in it) when you stop loading .torrent files unconditionally. It doesn't seem reliable to me. The user can upgrade from an older version for some reason. But most importantly, the user can have many paused torrents that do not overwrite their resume data.

It seems that this is a more difficult issue to solve along the way with this PR.

Reply:
>The user can upgrade from an older version for some reason. But most importantly, the user can have many paused torrents that do not overwrite their resume data.

Then that basically means we couldn't drop the .torrent loading code for a long time.

>It seems that this is a more difficult issue to solve along the way with this PR.

The problem you mentioned will **only** happen once we start dropping .torrent loading, while this PR in current form will work without problem for now and also solves the web seed issue.